homeassistant/components/bluetooth/active_update_coordinator.py

```diff
@@ -0,0 +1,141 @@
+):
+    """
+    A coordinator that parses passive data from advertisements but can also poll.
+
+    Every time an advertisement is received, needs_poll_method is called to work
+    out if a poll is needed. This should return True if it is and False if it is
+    not needed.
+
+    def needs_poll_method(svc_info: BluetoothServiceInfoBleak, last_poll: float | None) -> bool:
+        return True
+
+    If there has been no poll since HA started, `last_poll` will be None. Otherwise it is
+    the number of seconds since one was last attempted.
+
+    If a poll is needed, the coordinator will call poll_method. This is a coroutine.
+    It should return the same type of data as your update_method. The expectation is that
+    data from advertisements and from polling are being parsed and fed into a shared
+    object that represents the current state of the device.
+
+    async def poll_method(svc_info: BluetoothServiceInfoBleak) -> YourDataType:
+        return YourDataType(....)
+
+    BluetoothServiceInfoBleak.device contains a BLEDevice. You should use this in
+    your poll function, as it is the most efficient way to get a BleakClient.
+    """
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        logger: logging.Logger,
+        *,
+        address: str,
+        mode: BluetoothScanningMode,
+        update_method: Callable[[BluetoothServiceInfoBleak], _T],
+        needs_poll_method: Callable[[BluetoothServiceInfoBleak, float | None], bool],
+        poll_method: Callable[
+            [BluetoothServiceInfoBleak],
+            Coroutine[Any, Any, _T],
+        ]
+        | None = None,
+        poll_debouncer: Debouncer[Coroutine[Any, Any, None]] | None = None,
+    ) -> None:
+        """Initialize the processor."""
+        super().__init__(hass, logger, address, mode, update_method)
+
+        self._needs_poll_method = needs_poll_method
+        self._poll_method = poll_method
+        self._last_poll: float | None = None
+        self.last_poll_successful = True
+
+        # We keep the last service info in case the poller needs to refer to
+        # e.g. its BLEDevice
+        self._last_service_info: BluetoothServiceInfoBleak | None = None
+
+        if poll_debouncer is None:
+            poll_debouncer = Debouncer(
+                hass,
+                logger,
+                cooldown=POLL_DEFAULT_COOLDOWN,
+                immediate=POLL_DEFAULT_IMMEDIATE,
+                function=self._async_poll,
+            )
+        else:
+            poll_debouncer.function = self._async_poll
+
+        self._debounced_poll = poll_debouncer
+
+    def needs_poll(self, service_info: BluetoothServiceInfoBleak) -> bool:
+        """Return true if time to try and poll."""
+        poll_age: float | None = None
+        if self._last_poll:
+            poll_age = time.monotonic() - self._last_poll
+        return self._needs_poll_method(service_info, poll_age)
+
+    async def _async_update_data(
```

Comment:
Maybe rename to go with `poll_method`?
```suggestion
    async def _async_poll_data(
```